#123 Implement memory footprint mode

cz4rs opened the pull request:
fixes #123

cz4rs committed:
#123 add README.md

cz4rs committed:
#123 implement memory footprint mode

Implement a function and a mode for measuring total memory footprint of
an object.

- use existing 'serialize' infrastructure
- add implementation and tests for std::unique_ptr, std::string, std::vector

cz4rs commented:
note:
pushing some code here, but it's definitely still work-in-progress (requires improved naming (!) and docs, reduced code duplication, independence from `serialize` / separate operator, implementation for other containers etc.)

cz4rs committed:
#123 implement memory footprinting for standard containers

- adjust serialization code for Footprinting mode
- add unit tests for measuring memory footprint of standard containers
- fix compiler warning in traversal test

cz4rs committed:
#123 make CI tests output more verbose

cz4rs commented:
@lifflander I have removed the custom dispatcher and used existing `serialize` code when possible, can you have a look?

- standard containers, fundamental types, enums seem to be working fine, code for `std::tuple`, `std::function` and other necessary types will land tomorrow
- the tests are failing currently because of how raw pointers are handled (currently just counting the pointer size and following it, but that doesn't work for C-style arrays and C strings) - any idea how to proceed with this?
- I have updated https://github.com/DARMA-tasking/vt/pull/1013 to have an example on the `vt` side, if that looks ok I will start working on components one by one to see if any issues come out
- how much testing is required for when `detector` is not available? (currently the CI only tests with `detector` present)

cz4rs committed:
#123 implement memory footprinting for std::function

cz4rs committed:
#123 implement memory footprinting for std::queue and std::priority_queue

cz4rs committed:
#123 implement memory footprinting for std::chrono::duration

cz4rs committed:
#123 implement memory footprinting for std::shared_ptr

cz4rs committed:
#123 implement memory footprinting for FILE

cz4rs committed:
#123 compare calculated footprint with sizeof

PhilMiller commented on src/checkpoint/container/array_serialize.h:
```diff
@@ -54,6 +54,7 @@ namespace checkpoint {
 
 template <typename Serializer, typename T, size_t N>
 void serialize(Serializer& s, std::array<T, N>& array) {
+  s.countBytes(array);
   dispatch::serializeArray(s, &array[0], array.size());
 }
```
I think this is double-counting, since the implementation is required to contain the `T[N]` as a member, not point to it. So, I think the `countBytes` can just be omitted

cz4rs replied:
removed the redundant `countBytes` call

PhilMiller commented on src/checkpoint/container/chrono_serialize.h:
```diff
@@ -0,0 +1,62 @@
+
+template <typename Serializer, typename Rep, typename Period>
+void serialize(Serializer& s, std::chrono::duration<Rep, Period> d) {
+  s.countBytes(d);
```
Since this can only compute a footprint, it should probably assert that the serializer is in that mode, rather than silently failing

lifflander replied:
For this one, can we add a real serializer also here? I don't think chrono should be that difficult to serialize

cz4rs replied:
serializer added

PhilMiller commented on src/checkpoint/container/queue_serialize.h:
```diff
@@ -0,0 +1,79 @@
+  while (not q.empty())
+  {
+    s | q.front();
+    q.pop();
```
These implementations are destructive, which is quite undesirable.

cz4rs replied:
I couldn't find a better (concise) way to iterate over queue's elements, so I went with passing by value - `pop` is called on the copy. This is not too elegant and potentially slow, but the original object remains untouched.

lifflander replied:
It's quite unfortunate that this container (queue) doesn't have access to the underlying data structure.

PhilMiller commented on src/checkpoint/container/raw_ptr_serialize.h:
```diff
@@ -0,0 +1,81 @@
+ */
+template <typename Serializer, typename T>
+void serialize(Serializer& s, T* ptr) {
+  s.countBytes(ptr);
```
Again, since this will only work for footprinting, it should assert `s.isFootprinting()`

cz4rs replied:
enforcing Footprinting mode with `std::enable_if_t`

PhilMiller commented on src/checkpoint/container/shared_ptr_serialize.h:
```diff
@@ -0,0 +1,63 @@
+namespace checkpoint {
+
+template <typename Serializer, typename T>
+void serialize(Serializer& s, std::shared_ptr<T>& ptr) {
```
Ditto about assertion

cz4rs replied:
`std_enable_if_t` added

PhilMiller commented on src/checkpoint/container/string_serialize.h:
```diff
@@ -61,8 +61,13 @@ void serializeStringMeta(Serializer& s, std::string& str) {
-  dispatch::serializeArray(s, str.c_str(), str.size());
+  if (s.isFootprinting()) {
+    s.countBytes(str);
+    dispatch::serializeArray(s, str.c_str(), str.capacity());
```
This too potentially runs into the small-array optimization, but I think it's *not* a problem here, because `std::string` has a *fixed* small array size, so the overcount is tightly limited

PhilMiller commented:
Overall, I like this.

We'll want to figure out some systematic way to identify whether a class needs special footprint handling, and whether we've considered the need for special handling in each class for which we have a serializer.

Since we're handling container classes of various sorts in checkpoint, the serializers for `Kokkos` types need to be considered, too. I'm a little concerned that we don't correctly handle the special case of `Kokkos::View<Kokkos::View<T*>>` - the nested views will take up space even if empty, and need to be counted.

cz4rs committed:
#123 avoid double-counting footprint for std::array

cz4rs committed:
#123 prohibit use in other modes whe only Footprinting mode is supported

cz4rs commented on src/checkpoint/checkpoint_api.impl.h:
```diff
@@ -91,6 +91,12 @@ std::size_t getSize(T& target) {
 
+template <typename T>
+std::size_t getMemoryFootprint(T& target) {
+  // FIXME: use std::max to make sure that it's not smaller than sizeof(T)
```
at this stage (of implementation) `getMemoryFootprint` sometimes returns smaller value than a simple `sizeof`
(this can be valid calculation because of padding / layout)

note to self: change to `std::max(footprint, sizeof(T))` before merging

lifflander commented on src/checkpoint/container/raw_ptr_serialize.h:
```diff
@@ -0,0 +1,89 @@
+ * \param[in] ptr pointer to serialize
+ */
+template <typename Serializer, typename T>
+void serialize(Serializer& s, T* ptr) {
```
I think we need to bifurcate the dispatch for non-serializable, but footprintable types. Or cases where we want footprinting to be handled quite differently than regular serialization  We could do that statically, with an enable_if on the `SerializerT`. See comment at the bottom.

lifflander commented on src/checkpoint/container/raw_ptr_serialize.h:
```diff
@@ -0,0 +1,89 @@
+
+template <typename Serializer>
+void serialize(Serializer& s, FILE* ptr) {
+  checkpointAssert(
```
Same here.

lifflander commented:
@cz4rs @PhilMiller I'd like to include enable_if's on footprinting-only serializers, so that they can't accidentally be used by other serializers. This also makes it clear why the code doesn't compile if some footprint specific trait (like `s.countBytes`) is called and then fails to work with Packer/Unpacker/Sizer.

```C++
template <
  typename SerializerT,
  typename T,
  typename = std::enable_if_t<std::is_same<SerializerT, checkpoint::Footprinter>::value>
>
void serialize(SerializerT& s, std::queue<T> queue) {

}
```

cz4rs commented:
> I'd like to include enable_if's on footprinting-only serializers, so that they can't accidentally be used by other serializers. This also makes it clear why the code doesn't compile if some footprint specific member (like `s.countBytes`) is called and then fails to work with Packer/Unpacker/Sizer.

I have implemented `enable_if`'s in a helper function called by `serialize` and using anything else than `getMemoryFootprint` fails during compilation as expected. Putting `enable_if` directly in `serialize` was causing problems with the detection.

cz4rs commented on src/checkpoint/container/raw_ptr_serialize.h:
```diff
@@ -0,0 +1,116 @@
+ * \brief Serialize raw pointer \c ptr
+ *
+ * Only footprinting mode is supported at the moment. Counts the pointer size
+ * and follows it (note that it doesn't work correctly for C-style arrays!).
```
handling arrays correctly could be possible using `std::is_array` or
```cpp
template<typename SerializerT, typename T, size_t N> 
void serialize(T (&t)[N])
```

I poked around it a little bit, but currently arrays decay into pointers in the middle of call stack and I have left the warning here

cz4rs committed:
#123 complete serialization support for std::chrono::duration

PhilMiller commented on src/checkpoint/container/queue_serialize.h:
```diff
@@ -0,0 +1,104 @@
+    >::value
+  >
+>
+void serializeQueue(SerializerT& s, std::queue<T> q) {
```
Per discussion on today's call, we prefer to just count `q.size() * sizeof(T)`, and sacrifice some accuracy to avoid the potential pitfall of copying and iterating over every element.

PhilMiller replied:
That would shift us back to taking the queue arguments by `const&`

cz4rs replied:
done, the inaccuracy will be bigger when the underlying container is `std::vector` (default for priority queue), but there's no access to capacity in that case

PhilMiller replied:
I realize that for the specific cases in the runtime of these containers, we can handle footprinting specially, by keeping a running measure of the footprint of the contents, to add to the footprint explicitly.

Which means, the footprint interface needs to be able to accept "add arbitrary byte count"

PhilMiller commented on src/checkpoint/container/raw_ptr_serialize.h:
```diff
@@ -0,0 +1,116 @@
+void serializeFilePtr(SerializerT& s, FILE* ptr) {
+  s.countBytes(ptr);
+  if (ptr != nullptr) {
+    s.countBytes(*ptr);
```
On some platforms, I believe `struct FILE;` is left as an incomplete definition, to avoid exposing the C library's implementation. I think this would fail to compile in those settings, since it's ultimately going to attempt `sizeof()` on an incomplete type.

cz4rs replied:
changed to size the pointer only

PhilMiller commented on src/checkpoint/container/string_serialize.h:
```diff
@@ -59,10 +58,25 @@ void serializeStringMeta(Serializer& s, std::string& str) {
+/**
+ * \brief Serialize string \c str
+ *
+ * Resizes a string to it's actual size and serializes it.
```
Typo - should be "its"

cz4rs replied:
fixed

cz4rs committed:
#123 fix code review issues

- do not size underlying 'struct FILE' when footprinting FILE* pointer
- simplify queue serialization to avoid copying
- fix grammar typo

cz4rs committed:
#123 make sure that memory footprint is not smaller than 'sizeof'

cz4rs commented:
@PhilMiller @lifflander if this looks ok for you, I would like to get it merged to have better CI feedback in https://github.com/DARMA-tasking/vt/pull/1013

(I would be adding support for kokkos types and ironing out any discovered bugs in a subsequent PR)

cz4rs committed:
#123 implement memory footprinting for std::vector<bool>

cz4rs commented:
@lifflander @PhilMiller (thanks for the review!)
regarding the merge permissions / protected branches: please take a look at the screenshot below, I cannot merge this pull request myself
- the build passed fine
- review is approved "by reviewer with write access"
- the "not authorized" comment links to https://docs.github.com/en/github/administering-a-repository/about-protected-branches

I don't have a fresh example from `vt` repository, but it looked pretty similar there (we discussed this some time ago on DARMA technical meeting)

![image](https://user-images.githubusercontent.com/4639165/93398039-124e2b00-f87b-11ea-9f1f-e67cf432f163.png)

lifflander commented:
> @lifflander @PhilMiller (thanks for the review!)
> regarding the merge permissions / protected branches: please take a look at the screenshot below, I cannot merge this pull request myself
> 
> * the build passed fine
> * review is approved "by reviewer with write access"
> * the "not authorized" comment links to https://docs.github.com/en/github/administering-a-repository/about-protected-branches
> 
> I don't have a fresh example from `vt` repository, but it looked pretty similar there (we discussed this some time ago on DARMA technical meeting)
> 
> ![image](https://user-images.githubusercontent.com/4639165/93398039-124e2b00-f87b-11ea-9f1f-e67cf432f163.png)

I believe that I have fixed this now for this repository.
